File: Videos/YoutubePlaylistsDownload.py
#!/usr/bin/env python3

import datetime
import json
import math
import os
import re
import shutil
import subprocess
import sys
import time

if not os.path.exists("playlists.json"):
    raise FileNotFoundError("Missing playlists.json")
with open("playlists.json", "r") as f:
    playlists = json.load(f)
    playlists = dict(
        filter(
            lambda item: "disabled" not in item[1] or not item[1]["disabled"],
            playlists.items(),
        )
    )
    for playlist_name in playlists:
        if "url" not in playlists[playlist_name]:
            raise ValueError(
                f"Missing url for playlist {playlist_name} in playlists.json"
            )
        if "min_date" not in playlists[playlist_name]:
            playlists[playlist_name]["min_date"] = "1970-01-01"
        if "ids_desc_print_skip" not in playlists[playlist_name]:
            playlists[playlist_name]["ids_desc_print_skip"] = []
        if "ongoing" not in playlists[playlist_name]:
            playlists[playlist_name]["ongoing"] = False
        if "accurate" not in playlists[playlist_name]:
            playlists[playlist_name]["accurate"] = False
        if "path" not in playlists[playlist_name]:
            playlists[playlist_name]["path"] = playlist_name
        if "max_count" not in playlists[playlist_name]:
            playlists[playlist_name]["max_count"] = 100

dump_filename = ".dump.json"
videos_filename = ".videos.json"
downloading_folder = "Downloading"
bytes_per_second = 463_000
min_space_left = 2


def natural_sort(l, k=None):
    convert = lambda text: int(text) if text.isdigit() else text.lower()
    alphanum_key = lambda key: [
        convert(c) for c in re.split("([0-9]+)", key[k] if k else key)
    ]
    return sorted(l, key=alphanum_key)


def download_metadata(force=False):
    for playlist_name in playlists:
        path = playlists[playlist_name]["path"]
        if not os.path.exists(path):
            os.mkdir(path)
        dump_filepath = os.path.join(path, dump_filename)
        videos_filepath = os.path.join(path, videos_filename)

        if (
            not force
            and os.path.exists(dump_filepath)
            and os.path.exists(videos_filepath)
        ):
            modified_time = os.path.getmtime(videos_filepath)
            time_difference = time.time() - modified_time
            hours_difference = time_difference / 3600
            if hours_difference < 6:
                continue

        if (
            not playlists[playlist_name]["ongoing"]
            and os.path.exists(dump_filepath)
            and os.path.exists(videos_filepath)
            and not force
        ):
            continue

        print(f"Downloading metadata for {playlist_name}")
        with open(dump_filepath, "w") as f:
            ytdlp = subprocess.run(
                [
                    "yt-dlp",
                    "-J",
                    "--flat-playlist",
                    "--extractor-args",
                    "youtubetab:approximate_date",
                    playlists[playlist_name]["url"],
                ],
                stdout=f,
                stderr=subprocess.PIPE,
                text=True,
            )
            if ytdlp.returncode != 0:
                print(ytdlp.stderr)
                sys.exit(ytdlp.returncode)

        with open(dump_filepath, "r") as f:
            metadata = json.loads(f.read())

        videos = []
        atrioc_pattern = r"Streamed Live on (January|February|March|April|May|June|July|August|September|October|November|December) (\d{1,2})(?:.*?),? (\d{4})"
        for entry in metadata["entries"]:
            if playlist_name == "Atrioc":
                match = re.search(atrioc_pattern, entry["description"])
                if match:
                    month_name, day, year = match.groups()
                    date_object = datetime.datetime.strptime(
                        f"{month_name} {day} {year}", "%B %d %Y"
                    )
                else:
                    if (
                        entry["id"]
                        not in playlists[playlist_name]["ids_desc_print_skip"]
                    ):
                        print(f"No date found in the description for {entry['id']}")
                        print(entry["description"])
                    date_object = datetime.datetime.fromtimestamp(entry["timestamp"])
            else:
                # Private video
                if entry["timestamp"] is None:
                    continue
                date_object = datetime.datetime.fromtimestamp(entry["timestamp"])
            videos.append(
                {
                    "id": entry["id"],
                    "url": entry["url"],
                    "title": entry["title"],
                    "date": date_object.strftime("%Y-%m-%d"),
                }
            )
        if playlists[playlist_name]["ongoing"]:
            videos = sorted(videos, key=lambda x: x["date"], reverse=False)
        with open(videos_filepath, "w") as f:
            json.dump(videos, f, indent=4)


def download_videos():
    if not os.path.exists(downloading_folder):
        os.mkdir(downloading_folder)
    for playlist_name in playlists:
        printed_about_downloading_playlist = False
        path = playlists[playlist_name]["path"]
        if os.path.exists(os.path.join(path, ".archive")):
            with open(os.path.join(path, ".archive"), "r") as f:
                archive = [line.strip() for line in f.readlines()]
        else:
            archive = []
        with open(os.path.join(path, videos_filename), "r") as f:
            videos = json.loads(f.read())
        for i, video in enumerate(videos):
            download_count = len(
                list(filter(lambda f: f.endswith(".mp4"), os.listdir(path)))
            )
            if download_count >= playlists[playlist_name]["max_count"]:
                break
            if (
                video["date"] < playlists[playlist_name]["min_date"]
                or f"youtube {video['id']}" in archive
            ):
                continue

            if not printed_about_downloading_playlist:
                print(f"Downloading videos for {playlist_name}")
                printed_about_downloading_playlist = True

            print(f"Checking space for {video['title']}")
            space_left = compare_free_space_with_video(video["url"])
            if space_left < min_space_left:
                break
            output_name = f"%(title)s [%(id)s].%(ext)s"
            if playlists[playlist_name]["ongoing"]:
                output_name = f"{video['date']} - {output_name}"
            else:
                output_name = (
                    f"{i+1:0{math.floor(math.log10(len(videos))) + 1}} - {output_name}"
                )
            if playlists[playlist_name]["accurate"]:
                output_name = f"%(upload_date)s {output_name}"
            ytdlp = subprocess.run(
                [
                    "yt-dlp",
                    "-f",
                    "bestvideo[height<=1080][ext=mp4]+bestaudio[ext=m4a]/best[ext=mp4]/best",
                    "--downloader",
                    "aria2c",
                    "--download-archive",
                    (
                        os.path.join(path, ".archive")
                        if os.path.isabs(path)
                        else os.path.join("..", path, ".archive")
                    ),
                    "-o",
                    output_name,
                    video["url"],
                ],
                cwd=downloading_folder,
                check=True,
            )
            if ytdlp.returncode != 0:
                print(ytdlp.stderr)
                continue
            download_count += 1
            clean_downloading_folder(video["id"], path)
    shutil.rmtree(downloading_folder)


def clean_downloading_folder(id, copy_to, to_filename=None):
    for filename in os.listdir(downloading_folder):
        if filename.find(f"[{id}]"):
            if to_filename is None:
                to_filename = filename
            if filename.endswith(".mp4") or filename.endswith(".webm"):
                bps = find_bytes_per_second(os.path.join("Downloading", filename))
                if bps > bytes_per_second:
                    print(f"NEW LARGEST BYTES PER SECOND: {bps}")
                shutil.move(
                    os.path.join("Downloading", filename),
                    os.path.join(copy_to, to_filename),
                )
            else:
                os.remove(os.path.join("Downloading", filename))


def compare_free_space_with_video(video_url):
    free_space = shutil.disk_usage(".").free
    ytdlp = subprocess.run(
        ["yt-dlp", "--print", "%(filesize)s,%(duration)s", video_url],
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
        text=True,
    )
    if ytdlp.returncode != 0:
        print(ytdlp.stderr)
        return 0
    filesize, duration = ytdlp.stdout.split(",")
    try:
        filesize = int(filesize)
    except ValueError:
        filesize = int(duration) * bytes_per_second

    free_gbs = free_space / 1024.0 / 1024 / 1024
    video_gbs = int(filesize) / 1024.0 / 1024 / 1024
    print(f"{video_gbs:.2f} / {free_gbs:.2f} GB")
    return free_gbs - video_gbs


def find_bytes_per_second(filepath):
    probe = subprocess.run(
        [
            "ffprobe",
            "-v",
            "error",
            "-show_entries",
            "format=duration",
            "-of",
            "default=noprint_wrappers=1:nokey=1",
            filepath,
        ],
        text=True,
        check=True,
        stdout=subprocess.PIPE,
    )
    filesize = os.path.getsize(filepath)
    duration = float(probe.stdout)
    return filesize / duration


def find_all_bytes_per_second():
    arr = []
    for filename in os.listdir("Atrioc"):
        if filename.startswith("."):
            continue
        arr.append(find_bytes_per_second(os.path.join("Atrioc", filename)))
    arr = sorted(arr)
    print("\n".join(map(lambda x: str(int(x)), arr)))


def main():
    download_metadata()
    download_videos()


if __name__ == "__main__":
    if len(sys.argv) == 1:
        main()
    else:
        for arg in sys.argv[1:]:
            if arg == "download":
                download_videos()
            elif arg == "metadata":
                download_metadata()
            elif arg == "metadata-force":
                download_metadata(True)
            elif arg == "bps":
                find_all_bytes_per_second()
            elif arg == "help":
                print(
                    """
Usage:
    index.py <commands>

Commands:
    download
    metadata
    help"""
                )
            else:
                raise ValueError(f"Unknown argument {arg}")

File: Videos/YoutubePlaylistsDownload.ts
#!/usr/bin/env -S npx tsx

import { Option, program } from 'commander';
import { differenceInHours, format } from 'date-fns';
import { ensureDirSync, existsSync, moveSync, readdirSync, readFileSync, readJsonSync, rmSync, statfsSync, statSync, writeJsonSync } from 'fs-extra';
import { compact, padStart } from 'lodash';
import { resolve } from 'path';
import { z } from 'zod';
import { $ } from 'zx';
import { playlists as inputPlaylists } from './playlists';
import { Entry, Metadata, Playlist, Video } from './utils';

const DUMP_FILENAME = '.dump.json';
const VIDEOS_FILENAME = '.videos.json';
const ARCHIVE_FILENAME = '.archive';
const DOWNLOADING_FOLDER = 'Downloading';
const BYTES_PER_SECOND = 566_000;
const MIN_GBS_LEFT = 2;
const MIN_HOURS_TO_UPDATE_METADATA = 6;

const playlists = inputPlaylists as Record<string, Playlist>;
const schema = z.strictObject({
  url: z.string().url(),
  min_date: z.string().date().default('1970-01-01'),
  ongoing: z.boolean().default(false),
  accurate: z.boolean().default(false),
  path: z.string().default(''),
  max_count: z.number().int().positive().default(100),
  disabled: z.boolean().default(false),
  getEntry: z.function(z.tuple([z.any()]), z.any()).optional(),
});
for (const playlistName in playlists) {
  const parsed = schema.parse(playlists[playlistName]);
  parsed.path = parsed.path || playlistName;
  if (parsed.disabled) delete playlists[playlistName];
  else playlists[playlistName] = parsed;
}

function freeDiskGBs() {
  const stat = statfsSync(DOWNLOADING_FOLDER);
  return stat.bavail * stat.bsize / 1024 / 1024 / 1024;
}

function fileGBs(path: string) {
  const stat = statSync(path);
  if (stat.isFile()) return stat.size / 1024 / 1024 / 1024;
  const files = readdirSync(path);
  let size = 0;
  for (const file of files) {
    size += fileGBs(resolve(path, file));
  }
  return size;
}

function videoGBs(videoUrl: string) {
  const $ytdlp = $({ sync: true, stdio: ['ignore', 'pipe', 'ignore'] })`yt-dlp --print "%(filesize)s,%(duration)s" ${videoUrl}`;
  if ($ytdlp.exitCode !== 0) return Number.POSITIVE_INFINITY;
  const [filesizeStr, durationStr] = $ytdlp.stdout.split(',');
  let filesize = +filesizeStr;
  if (isNaN(filesize)) filesize = +durationStr * BYTES_PER_SECOND;
  if (!filesize) return Number.POSITIVE_INFINITY;
  return filesize / 1024 / 1024 / 1024;
}

function downloadVideos(skipSizeCheck = false) {
  ensureDirSync(DOWNLOADING_FOLDER);
  for (const playlistName in playlists) {
    const playlist = playlists[playlistName];
    const videos: Video[] = readJsonSync(resolve(playlist.path, VIDEOS_FILENAME));
    let printedAboutDownloadingPlaylist = false;
    let archive: string[];
    try {
      archive = readFileSync(resolve(playlist.path, ARCHIVE_FILENAME), { encoding: 'utf-8' }).split(/\r?\n/);
    } catch (error) {
      archive = [];
    }

    for (let i = 0; i < videos.length; i++) {
      const video = videos[i];
      const downloadCount = readdirSync(playlist.path).filter(f => isFileVideo(f)).length;
      if (downloadCount >= playlist.max_count) break;
      if (video.date.localeCompare(playlist.min_date) < 0 || archive.includes(`youtube ${video.id}`)) continue;

      if (!printedAboutDownloadingPlaylist) {
        console.log(`Downloading videos for ${playlistName}`);
        printedAboutDownloadingPlaylist = true;
      }

      if (!skipSizeCheck) {
        console.log(`Checking space for ${video.title}`);
        const diskGBs = freeDiskGBs();
        const videoSize = videoGBs(video.url);
        const downloadingFolderSize = fileGBs(DOWNLOADING_FOLDER);
        console.log(`(${videoSize.toFixed(2)} * 2) / (${diskGBs.toFixed(2)} + ${downloadingFolderSize.toFixed(2)})`);
        if (diskGBs + downloadingFolderSize - videoSize * 2 < MIN_GBS_LEFT) break;
      }

      let outputName = '%(title)s [%(id)s].%(ext)s';
      if (playlist.ongoing) outputName = `${video.date} - ${outputName}`;
      else {
        const playlistIndex = i + 1;
        const padLength = Math.floor(Math.log10(videos.length)) + 1;
        outputName = `${padStart(playlistIndex.toString(), padLength, '0')} - ${outputName}`;
      }
      if (playlist.accurate) outputName = `%(upload_date)s ${outputName}`;

      const flags = [
        '-f',
        'bestvideo[height<=1080][ext=mp4]+bestaudio[ext=m4a]/best[ext=mp4]/best/bestvideo+bestaudio',
        '--downloader',
        'aria2c',
        '--download-archive',
        resolve(playlist.path, ARCHIVE_FILENAME),
        '-o',
        outputName,
        video['url'],
      ];
      const $ytdlp = $({ sync: true, cwd: DOWNLOADING_FOLDER, stdio: 'inherit' })`yt-dlp ${flags}`;
      if ($ytdlp.exitCode === 0) cleanDownloadingFolder(video.id, playlist.path);
    }
  }
}

function cleanDownloadingFolder(id: string, copyToDir: string) {
  const files = readdirSync(DOWNLOADING_FOLDER);
  for (const file of files) {
    if (file.indexOf(`[${id}]`) === -1) continue;
    if (isFileVideo(file)) {
      const bps = videoBPS(resolve(DOWNLOADING_FOLDER, file));
      if (bps > BYTES_PER_SECOND) {
        console.log(`NEW LARGEST BYTES PER SECOND: ${bps}`);
      }
      moveSync(resolve(DOWNLOADING_FOLDER, file), resolve(copyToDir, file));
    } else {
      rmSync(resolve(DOWNLOADING_FOLDER, file));
    }
  }
}

function printAllBPS(playlistName: string) {
  const bpss = [] as number[];
  const files = readdirSync(playlists[playlistName].path);
  for (const file of files) {
    if (!isFileVideo(file) || file.startsWith('.')) continue;
    bpss.push(videoBPS(resolve(playlists[playlistName].path, file)));
  }
  bpss.sort((a, b) => b - a);
  console.log(bpss.map(x => Math.floor(x)).join('\n'));
}

function videoBPS(filepath: string) {
  const $ffprobe = $({ sync: true, stdio: ['ignore', 'pipe', 'inherit'] })`ffprobe -v error -show_entries format=duration -of default=noprint_wrappers=1:nokey=1 ${filepath}`;
  const duration = +$ffprobe.stdout;
  const filesize = statSync(filepath).size;
  return filesize / duration;
}

function isFileVideo(filepath: string) {
  return filepath.endsWith('.mp4') || filepath.endsWith('.webm');
}

function getVideoFromEntry(playlist: Playlist, entry: Entry): Video | undefined {
  if (!entry.timestamp) return;

  let date = new Date(entry.timestamp * 1000);
  let title = entry.title;
  if (playlist.getEntry) {
    const data = playlist.getEntry(entry);
    if (data === null) {
      return;
    } else {
      if (data?.date) date = data.date;
      if (data?.title) title = data.title;
    }
  }
  return {
    id: entry.id,
    url: entry.url,
    date: format(date, 'yyyy-MM-dd'),
    title,
  };
}

function downloadMetadata(options: { force?: boolean, video?: boolean, skipSizeCheck: boolean }) {
  for (const playlistName in playlists) {
    const playlist = playlists[playlistName];
    const dumpFilepath = resolve(playlist.path, DUMP_FILENAME);
    const videosFilepath = resolve(playlist.path, VIDEOS_FILENAME);
    ensureDirSync(playlist.path);

    if (!options.force && existsSync(dumpFilepath) && existsSync(videosFilepath)) {
      const modifiedTime = statSync(videosFilepath).mtimeMs;
      const hours = differenceInHours(new Date(), new Date(modifiedTime));
      if (hours < MIN_HOURS_TO_UPDATE_METADATA) continue;
    }

    if (!playlist.ongoing && existsSync(dumpFilepath) && existsSync(videosFilepath) && !options.force)
      continue;

    if (!options.video || !existsSync(dumpFilepath)) {
      console.log(`Downloading metadata for ${playlistName}`);
      const tmpDumpFilepath = `${dumpFilepath}.tmp`;
      $({ sync: true })`yt-dlp -J --flat-playlist --extractor-args youtubetab:approximate_date ${playlist.url} > ${tmpDumpFilepath}`;
      moveSync(tmpDumpFilepath, dumpFilepath, { overwrite: true });
    }

    const metadata: Metadata = readJsonSync(dumpFilepath);
    const videos: Video[] = compact(metadata.entries.map(e => getVideoFromEntry(playlist, e)));
    if (playlist.ongoing) videos.sort((a, b) => a.date.localeCompare(b.date));
    writeJsonSync(videosFilepath, videos, { spaces: 2 });
  }
}

function main(options: { force?: boolean, skipSizeCheck: boolean }) {
  downloadMetadata(options);
  downloadVideos(options.skipSizeCheck);
}

program
  .nameFromFilename(__filename)
  .description('Download playlists listed in playlists.json');
program.command('run', { hidden: true, isDefault: true })
  .option('-f, --force', 'Force metadata update')
  .option('-s, --skip-size-check', 'Skip size check', false)
  .option('-p, --playlist <playlist_name>', 'S')
  .action(main);
program.command('download')
  .description('Download videos without updating metadata')
  .alias('d')
  .option('-s, --skip-size-check', 'Skip size check', false)
  .action(downloadVideos);
program.command('metadata')
  .description('Update metadata without downloading videos')
  .option('-f, --force', 'Force metadata update')
  .option('-s, --skip-size-check', 'Skip size check', false)
  .addOption(new Option('-v, --video', 'Update .videos.json without updating .dump.json').implies({ force: true }))
  .alias('m')
  .action(downloadMetadata);
program.command('bytes-per-second')
  .description('Find maximum bytes per second for a playlist')
  .alias('bps')
  .argument('[playlist-name]', 'Playlist to find max bps for', Object.keys(playlists)[0])
  .action(printAllBPS);
program.parse();
